### PR TITLE
The RemoteTerminal plugin 

### PR DESCRIPTION
**Pull request by marcin-zyrkowski**
Synchronized today (11.05) with the mendersoftware/mender-python-client/master.
Applied the requested changes on the reviews.
Unit tested, all 99 passed.
Tested on both virtual device (Ubuntu) as well on the Customer's device.
Please remember of squashing all development commits into the single one once merging.

### CODE REVIEW

File: src/mender/statemachine/statemachine.py [hunk 1]
```diff
@@ -173,7 +186,8 @@ def run(self, context):
                 )  # Update machine runs when idle detects an update
             except HTTPUnathorized:
                 context.authorized = False
-                return
+                return # @fixme can it be change to 'break'?
+        # @fixme: we are no longer authorized, the WebSocket connection needs to be closed
```
Comment (oleorhagen) [inline src/mender/statemachine/statemachine.py:189]: What's the status on these?

Comment (oleorhagen) [reply src/mender/statemachine/statemachine.py:189]: @marcin-zyrkowski

File: src/mender/remoteterminal/remoteterminal.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+import subprocess
+import threading
+
+import msgpack  # type: ignore
+import websockets
+from websockets.exceptions import WebSocketException
+
+log = logging.getLogger(__name__)
+
+# mender-connect protocol messeges, taken from:
```
Comment (oleorhagen) [inline src/mender/remoteterminal/remoteterminal.py:30]: s/messeges/messages/

File: src/mender/remoteterminal/remoteterminal.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        thread_ws = threading.Thread(
+            target=lambda: asyncio.run(self.proto_msg_processor())
+        )
+        thread_ws.start()
+
+    def open_terminal(self):
+        """opens new pty/tty, invokes the new shell and connects them"""
+        self.master, self.slave = pty.openpty()
+        # by default the shell owner is root
+        self.shell = subprocess.Popen(
```
Comment (oleorhagen) [inline src/mender/remoteterminal/remoteterminal.py:182]: Will this not leak zombies? 
It seems to me that it's only closed in the case of a `STOP` from the server :thinking:

**Review by oleorhagen — COMMENTED** [review]
Looks pretty I think :smile: 

I added a commit fixing up the static checks, and left a couple of comments in the code.

Good job @marcin-zyrkowski !

File: src/mender/statemachine/statemachine.py (hunk 1 above)
Comment (marcin-zyrkowski) [reply src/mender/statemachine/statemachine.py:189]: Line 189: At time I was reading and learning the code I spotted that 'break' would do the same as 'return' in this place. I prefer an approach of "single exit/return point from the function" so I placed the comment to get back to this later. However this is not the time for refactoring now and I'm happy to reject my comment and leave the line as it was.
Line 190: There is no way of unauthorizing the device during the remote session as the term session in Web UI is modal and not allow to exit without closing the session. So this comment is invalid and shout be removed.

Comment (marcin-zyrkowski) [reply src/mender/statemachine/statemachine.py:189]: I will remove both comments.

File: src/mender/remoteterminal/remoteterminal.py (hunk 2 above)
Comment (marcin-zyrkowski) [reply src/mender/remoteterminal/remoteterminal.py:30]: Will fix.

File: src/mender/remoteterminal/remoteterminal.py (hunk 3 above)
Comment (marcin-zyrkowski) [reply src/mender/remoteterminal/remoteterminal.py:182]: I agree that better solution would be to self-cleaning however the many tests showed that Python (maybe via destruction of the object returned by Popen) closes that links. In fact even when the method self.shell.kill() is invoked the defunc process (zombie) lives in system - but it is cleaned internally when another session is started and it dies along with the parent when daemon is killed.

**Comment by oleorhagen** [comment]
Closed in favour of: https://github.com/mendersoftware/mender-python-client/pull/79

**Pull request closed without merging**